aio.core/aio/core/functional/generator.py

```diff
@@ -0,0 +1,62 @@
+
+from typing import AsyncIterable, AsyncGenerator, Callable, Coroutine, Iterable
+
+from aio.core import functional
+
+
+class AwaitableGenerator:
+    """Wrap an asynchronous generator as optionally awaitable.
+
+    For example, you may have a low level async function that generates
+    results, but where in some cases you want to just wait for the gathered
+    results.
+
+    The 2 methods of handling results show below are functionally equivalent,
+    save for the that the in the second example all results are generated
+    before iteration begins.
+
+    ```python
+
+    async def some_generator():
+        for x in range(0, 5):
+            yield x
+
+    async for result in AwaitableGenerator(some_generator()):
+        do_something_with(result)
+
+    for result in await AwaitableGenerator(some_generator()):
+        do_something_with(result)
+
+    ```
+
+    By default results will be collected into a `list`, using the
+    `async_list` function. You can specify a custom collector as
+    an argument to the constructor.
+
+    Any additional kwargs passed to the constructor are passed on to the
+    collector, but are not used during normal generation.
```

Comment:
maybe they should be - i think for this interface to make sense it needs to return the same results by either method